### PR TITLE
many: add reseal API/command

### PR DESCRIPTION
**Pull request by valentindavid**

### CODE REVIEW

File: overlord/devicestate/devicestate.go [hunk 1]
```diff
@@ -1479,3 +1479,10 @@ func InstallSetupStorageEncryption(st *state.State, label string, onVolumes map[
 
 	return chg, nil
 }
+
+func Reseal(st *state.State) *state.Change {
+	chg := st.NewChange("reseal-device", i18n.G("Reseal the device"))
```
Comment (bboozzoo) [inline overlord/devicestate/devicestate.go:1484]: The change may need to be accounted for in conflicts. I suppose for simplicity we may disallow any other changes from running, cc @pedronis

Comment (valentindavid) [reply overlord/devicestate/devicestate.go:1484]: Out of curiosity, what conflict are you thinking of?

Comment (bboozzoo) [reply overlord/devicestate/devicestate.go:1484]: Wondering if there could be a potential issue if the there is a create recovery system create or remodel or gadget update. IIRC the remodel and recovery wee made exclusive in https://github.com/snapcore/snapd/blob/master/overlord/snapstate/conflict.go#L147 not sure if that has changed.

**Comment by valentindavid** [comment]
@pedronis this is the minimal implementation. I have only done manual tests. I will write tests now. But it would be nice if you could go through this minimal change and tell me if things should be moved, or if there is anything missing (like for instance what @bboozzoo suggested).

File: boot/seal.go [hunk 2]
```diff
@@ -472,7 +472,7 @@ func resealKeyToModeenvUsingFDESetupHookImpl(rootdir string, modeenv *Modeenv, e
 }
 
 // TODO:UC20: allow more than one model to accommodate the remodel scenario
-func resealKeyToModeenvSecboot(rootdir string, modeenv *Modeenv, expectReseal bool) error {
+func resealKeyToModeenvSecboot(rootdir string, modeenv *Modeenv, expectReseal, forceReseal bool) error {
```
Comment (pedronis) [inline boot/seal.go:475]: I would bundle the two flags into a struct

File: overlord/devicestate/devicemgr.go [hunk 3]
```diff
@@ -214,6 +214,8 @@ func Manager(s *state.State, hookManager *hookstate.HookManager, runner *state.T
 
 	runner.AddBlocked(gadgetUpdateBlocked)
 
+	runner.AddHandler("reseal", m.doReseal, nil)
```
Comment (pedronis) [inline overlord/devicestate/devicemgr.go:217]: maybe boot-reseal

File: overlord/devicestate/devicestate.go (hunk 1 above)
Comment (pedronis) [inline overlord/devicestate/devicestate.go:1484]: Reseal device against boot parameters

File: boot/force_reseal.go [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+ *
+ */
+
+package boot
+
+import (
+	"github.com/snapcore/snapd/dirs"
+)
+
+func ForceReseal(unlocker Unlocker) error {
```
Comment (pedronis) [inline boot/force_reseal.go:26]: needs a doc comment

File: daemon/api_reseal.go [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+import (
+	"net/http"
+
+	"github.com/snapcore/snapd/overlord/auth"
+	"github.com/snapcore/snapd/overlord/devicestate"
+)
+
+var sealCmd = &Command{
+	Path:        "/v2/reseal",
```
Comment (pedronis) [inline daemon/api_reseal.go:30]: maybe /v2/system-reseal like /v2/sysem-recover-keys or a /v2/systems POST action ?

File: cmd/snap/cmd_reseal.go [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+TODO
+`)
+)
+
+type cmdReseal struct {
+	waitMixin
+}
+
+func init() {
+	addCommand("reseal",
```
Comment (pedronis) [inline cmd/snap/cmd_reseal.go:40]: as it's related to booting/system I would perhaps make it a subcommand of "snap reboot":

snap reboot --reseal

we have a few subcommands attached that way to other commands

**Review by pedronis — COMMENTED** [review]
thank you, some initial considerations and small comments

File: daemon/api_reseal.go (hunk 5 above)
Comment (valentindavid) [reply daemon/api_reseal.go:30]: I will go for the first option now. Unless there is a meaning behind resealing with a specify system label?

**Comment by valentindavid** [comment]
I have managed to get it to work on classic. However there are some manual steps required:
 * Boot with recovery keys
 * Remove `/var/lib/snapd/save/device/fde/tpm-lockout-auth`
 * Write `5` to `/sys/class/tpm/tpm0/ppi/request`.
 * Reboot, confirm the reset, then unlock with recovery keys
 * Run `snap reboot --reset`, it will reboot
 * Your system is fixed.

@pedronis I wonder if we can make changes across multiple boot. That is request a reboot and then continue resetting, then request another reboot.

**Comment by valentindavid** [comment]
It feels to me that the locking of tpm when booting with recovery keys is a bug. We should not try to call `MarkBootSuccess` when we have used recovery keys.

File: cmd/snap/cmd_reseal.go (hunk 6 above)
Comment (valentindavid) [reply cmd/snap/cmd_reseal.go:40]: I have not yet removed this command. I have added  `snap reboot --reseal` however.

**Comment by valentindavid** [comment]
Marking as blocked because of lockout behavior on fresh TPM.

**Comment by valentindavid** [comment]
Closed because the same reason.

**Pull request closed without merging**